### PR TITLE
Thriftmux support

### PR DESCRIPTION
**Pull request by apakulov**
Had to use StackBasedServer trait instead of StackServer in protocol ProtocolInitializer to be compatible with ThriftMux

### CODE REVIEW

**Comment by apakulov** [comment]
@olix0r @adleong is there a chance to get your feedback on this PR? :-)

**Comment by olix0r** [comment]
@apakulov Please excuse the delay. We've been prioritizing issues which are blocking the 0.9.0 release. We'll try to get to this soon, though. Thanks for your patience.

**Comment by apakulov** [comment]
@olix0r completely understandable, thanks for the update.

File: linkerd/protocol/mux/src/main/resources/META-INF/services/io.buoyant.linkerd.ProtocolInitializer [hunk 1]
```diff
@@ -1 +0,0 @@
-io.buoyant.linkerd.protocol.MuxInitializer
```
Comment (apakulov) [inline linkerd/protocol/mux/src/main/resources/META-INF/services/io.buoyant.linkerd.ProtocolInitializer]: My bad, will bring it back.

File: linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxTraceInitializer.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package com.twitter.finagle.buoyant.linkerd
+
+import com.twitter.finagle._
+import com.twitter.finagle.mux.{Request, Response}
+import com.twitter.finagle.tracing.{Trace, TraceInitializerFilter, Tracer}
+import com.twitter.util.Future
+
+object ThriftMuxTraceInitializer {
```
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxTraceInitializer.scala]: Done.

File: linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala [hunk 3]
```diff
@@ -0,0 +1,46 @@
+package io.buoyant.linkerd
+package protocol
+
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.twitter.finagle.{Path, Stack}
+import com.twitter.finagle.buoyant.linkerd.ThriftMuxTraceInitializer
+import io.buoyant.router.{RoutingFactory, Thrift, ThriftMux}
+
+class ThriftMuxInitializer extends ProtocolInitializer {
```
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala]: Done.

File: linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  val name = "thriftmux"
+
+  protected type RouterReq = com.twitter.finagle.thrift.ThriftClientRequest
+  protected type RouterRsp = Array[Byte]
+  protected type ServerReq = Array[Byte]
+  protected type ServerRsp = Array[Byte]
+
+  protected val defaultRouter = ThriftMux.router
```
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala]: Done.

File: linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    .configured(RoutingFactory.DstPrefix(Path.Utf8(name)))
```
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala]: Removed.

File: linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  override def defaultServerPort: Int = 4144
+
+  val configClass = classOf[ThriftMuxConfig]
+}
+
+object ThriftMuxInitializer extends ThriftMuxInitializer
+
+class ThriftMuxConfig(thriftMethodInDst: Option[Boolean]) extends RouterConfig {
```
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/io/buoyant/linkerd/protocol/ThriftMuxInitializer.scala]: Using ``ThriftConfig`` as ``configClass``

**Comment by apakulov** [comment]
@adleong will add example, docs and entry CHANGES.md soon.

**Comment by apakulov** [comment]
@adleong ready for another review round.

File: linkerd/docs/protocol-thriftmux.md [hunk 7]
```diff
@@ -0,0 +1,36 @@
+# ThriftMux Protocol (experimental)
+
+> This config routes thriftmux from port 4400 to port 5005. 
+
+```yaml
+routers:
+- protocol: thriftmux
```
Comment (adleong) [inline linkerd/docs/protocol-thriftmux.md]: I think this config requires the experimental property to be set

File: linkerd/examples/thriftmux.yaml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+
+# An example service discovery that backend uses the filesystem to resolve endpoints.
+namers:
+- kind: io.l5d.fs
+  prefix: /thriftmux
+  rootDir: linkerd/examples/io.l5d.fs
+
+# A simple thriftmux router that looks up each thriftmux method in io.l5d.fs.
+routers:
+- protocol: thriftmux
```
Comment (adleong) [inline linkerd/examples/thriftmux.yaml]: I don't think this example routes correctly, but it looks like neither does the thrift.yaml example.  Perhaps we can just fix both of them in a followup review.

File: linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    ): ServiceFactory[Request, Response] = {
+      val arrayBytesServiceFactory = ArrayBytesToMuxFilter.andThen(next)
+      MuxToArrayBytesFilter.andThen(ThriftServerPrep.module.make(params, arrayBytesServiceFactory))
+    }
+  }
+
+  object MuxToArrayBytesFilter extends Filter[mux.Request, mux.Response, Array[Byte], Array[Byte]] {
+    def apply(request: mux.Request, service: Service[Array[Byte], Array[Byte]]): Future[mux.Response] = {
+      val reqBytes = Buf.ByteArray.Owned.extract(request.body)
+      service(reqBytes) map { repBytes =>
```
Comment (adleong) [inline linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala]: Tiny style nit: we usually prefer `service(reqBytes).map {` over `service(reqBytes) map {`

File: linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+        mux.Response(Buf.ByteArray.Owned(repBytes))
+      }
+    }
+  }
+
+  object ArrayBytesToMuxFilter extends Filter[Array[Byte], Array[Byte], mux.Request, mux.Response] {
+    def apply(request: Array[Byte], service: Service[mux.Request, mux.Response]): Future[Array[Byte]] = {
+      val req = mux.Request(Path.empty, Buf.ByteArray.Owned(request))
+      service(req) map { rep =>
```
Comment (adleong) [inline linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala]: here too

File: linkerd/docs/protocol-thriftmux.md (hunk 7 above)
Comment (apakulov) [inline linkerd/docs/protocol-thriftmux.md]: Added.

File: linkerd/examples/thriftmux.yaml (hunk 8 above)
Comment (apakulov) [inline linkerd/examples/thriftmux.yaml]: Done.

File: linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala (hunk 9 above)
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala]: Done.

File: linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala (hunk 10 above)
Comment (apakulov) [inline linkerd/protocol/thriftmux/src/main/scala/com/twitter/finagle/buoyant/linkerd/ThriftMuxServerPrep.scala]: Done.

**Pull request merged**